design/fast_finality.md

```diff
@@ -0,0 +1,273 @@
+  * Nodes can easily prove whether a given soft-finalised transaction was hard-finalised, and in the correct order
+* Value-extraction resistance
+  * The sequencer cannot precompute the effects of running a given set of transactions without committing to that set 
+    of transactions (e.g. by running a single transaction, then `eth_call`ing to see how it has affected a smart 
+    contract's state)
+* Cost
+  * L1 transaction costs can be driven lower at the expense of extending the hard-finality window
+  * The network can be configured to only publish rollups to the L1 every `x` blocks, where `x` >> 1
+  * The network can be configured to produce light batches (see below) at a higher frequency than L1 blocks
+* User/dev experience
+  * The responses to RPC calls reflect the soft-finalised transactions, and not just the hard-finalised transactions
+* Resilience
+  * Failover does not require a governance action
+  * During failover, it is acceptable to break the one-second soft-finality guarantee 
+  * During failover, it is acceptable to drop transactions that have not been soft-finalised
+
+## Assumptions
+
+* There is a single aggregator that is also the genesis node. This node is known as the *sequencer*
+* The enclave has a start-up delay that exceeds the production time for light batches (see below)
+
+## Design
+
+### Sequencer identity
+
+The sequencer's identity is given in the management contract on the L1 as a set of enclave attestations. This allows 
+other nodes to verify that the light batches and rollups are created by the sequencer. We use a set instead of a 
+single attestation to allow faster failover in the case of a sequencer node crashing (see the section `Resilience`, 
+below).
+
+Each attestation matches one of the sequencer's enclaves, and contains the hash of that enclave's key. Since the
+attestations are unique per machine, the enclaves cannot be impersonated. The foundation admin will then whitelist
+these attestations. If one of the sequencer's enclaves goes down irrecoverably, the old attestation can no longer be 
+used (since it is tied to the machine). A replacement attestation can then be added and whitelisted to induct a 
+replacement enclave into the sequencer's cluster, following the same process.
+
+### Production of light batches
+
+A light batch is produced on the required cadence to meet the network's soft-finality window of one second. Only the 
+sequencer produces light batches.
+
+To produce a light batch, the sequencer's host feeds a set of transactions to the enclave. The enclave responds by 
+creating a signed and encrypted *light batch*. This light batch is formally identical to the rollup of the final 
+design, including a list of the provided transactions and a header including information on the current light batch and 
+the hash of the "parent" light batch. This header also includes a monotonically increasing counter, in order to support 
```

Comment:
I would nuance this. 
I think we have to think a bit more about the right cadence of incrementing that counter that is fed to the smart contracts.
It could be the l1 height as well. I can't think now if it's necessarily a problem if the same counter is presented to the transactions in multiple batches.

Reply:
Cool - added to the unresolved issues for now.